some Go tests fail due to missing local temp directories

**Describe the bug**

When running the Go test suite on some development systems (macOS, at least) with `make test`, some tests fail with the errors of the form:
```
    testutils.go:228: Error running git command 'git push origin ...': exit status 1 EOF
        Uploading LFS objects: ...
        error: failed to push some refs to '/var/folders/.../T/lfsRepo...'
```

Running `make test` again with the `GIT_TRACE=1` environment variable yields the information that the given path plus `/.git` does not exist as a local directory.  As an example:
```
12:21:40.442232 trace git-lfs: xfer[git-lfs]: error creating handler: chdir /var/folders/f4/3tj6q1w11xd4mnnx7fm3jk8m0000gn/T/lfsRepo758791364/.git: no such file or directory
        EOF
        Uploading LFS objects: 100% (2/2), 45 B | 0 B/s, done.
        error: failed to push some refs to '/var/folders/f4/3tj6q1w11xd4mnnx7fm3jk8m0000gn/T/lfsRepo758791364'
```

**To Reproduce**
Steps to reproduce the behaviour:
1. Build from source per the Wiki [instructions](https://github.com/git-lfs/git-lfs/wiki/Installation#source).
2. Run the Go test suite with `make test` (not the shell tests in the `t/` directory).
3. Observe the `TestCurrentRefAndCurrentRemoteRef`, `TestRecentBranches`, and `TestScanUnpushed` tests fail with errors as described above.

**Expected behaviour**
All Go tests should pass.

**System environment**
macOS Catalina with Go version 1.13.3.

**Output of `git lfs env`**
Not applicable.

**Additional context**
Not applicable.

--- a/Makefile
+++ b/Makefile
@@ -406,7 +406,7 @@ test-race : GO_TEST_EXTRA_ARGS=-race
 # And so on.
 test : fmt
 	(unset GIT_DIR; unset GIT_WORK_TREE; \
-	$(GO) test $(GO_TEST_EXTRA_ARGS) $(addprefix ./,$(PKGS)))
+	$(GO) test -count=1 $(GO_TEST_EXTRA_ARGS) $(addprefix ./,$(PKGS)))
 
 # integration is a shorthand for running 'make' in the 't' directory.
 .PHONY : integration

--- a/Makefile
+++ b/Makefile
@@ -404,7 +404,7 @@ test-race : GO_TEST_EXTRA_ARGS=-race
 # 		make PKGS="config lfsapi tools/kv" test-race
 #
 # And so on.
-test : fmt
+test : fmt $(.DEFAULT_GOAL)
 	(unset GIT_DIR; unset GIT_WORK_TREE; \
 	$(GO) test -count=1 $(GO_TEST_EXTRA_ARGS) $(addprefix ./,$(PKGS)))
 

--- a/lfsapi/endpoint_finder_test.go
+++ b/lfsapi/endpoint_finder_test.go
@@ -1,6 +1,8 @@
 package lfsapi
 
 import (
+	"io/ioutil"
+	"os"
 	"runtime"
 	"testing"
 
@@ -313,32 +315,92 @@ func TestBareGitEndpointAddsLfsSuffix(t *testing.T) {
 	assert.Equal(t, "", e.SshPort)
 }
 
-func TestLocalPathEndpointAddsDotGitDir(t *testing.T) {
+func TestLocalPathEndpointAddsDotGitForWorkingRepo(t *testing.T) {
 	// Windows will add a drive letter to the paths below since we
 	// canonicalize them.
 	if runtime.GOOS == "windows" {
 		return
 	}
 
+	path, err := ioutil.TempDir("", "lfsRepo")
+	assert.Nil(t, err)
+	path = path + "/local/path"
+	err = os.MkdirAll(path+"/.git", 0755)
+	assert.Nil(t, err)
+
 	finder := NewEndpointFinder(lfshttp.NewContext(nil, nil, map[string]string{
-		"remote.origin.url": "/local/path",
+		"remote.origin.url": path,
 	}))
 	e := finder.Endpoint("download", "")
-	assert.Equal(t, "file:///local/path/.git", e.Url)
+	assert.Equal(t, "file://"+path+"/.git", e.Url)
+
+	os.RemoveAll(path)
 }
 
-func TestLocalPathEndpointPreservesDotGit(t *testing.T) {
+func TestLocalPathEndpointPreservesDotGitForWorkingRepo(t *testing.T) {
 	// Windows will add a drive letter to the paths below since we
 	// canonicalize them.
 	if runtime.GOOS == "windows" {
 		return
 	}
 
+	path, err := ioutil.TempDir("", "lfsRepo")
+	assert.Nil(t, err)
+	path = path + "/local/path/.git"
+	err = os.MkdirAll(path, 0755)
+	assert.Nil(t, err)
+
 	finder := NewEndpointFinder(lfshttp.NewContext(nil, nil, map[string]string{
-		"remote.origin.url": "/local/path.git",
+		"remote.origin.url": path,
 	}))
 	e := finder.Endpoint("download", "")
-	assert.Equal(t, "file:///local/path.git", e.Url)
+	assert.Equal(t, "file://"+path, e.Url)
+
+	os.RemoveAll(path)
+}
+
+func TestLocalPathEndpointPreservesNoDotGitForBareRepo(t *testing.T) {
+	// Windows will add a drive letter to the paths below since we
+	// canonicalize them.
+	if runtime.GOOS == "windows" {
+		return
+	}
+
+	path, err := ioutil.TempDir("", "lfsRepo")
+	assert.Nil(t, err)
+	path = path + "/local/path"
+	err = os.MkdirAll(path, 0755)
+	assert.Nil(t, err)
+
+	finder := NewEndpointFinder(lfshttp.NewContext(nil, nil, map[string]string{
+		"remote.origin.url": path,
+	}))
+	e := finder.Endpoint("download", "")
+	assert.Equal(t, "file://"+path, e.Url)
+
+	os.RemoveAll(path)
+}
+
+func TestLocalPathEndpointRemovesDotGitForBareRepo(t *testing.T) {
+	// Windows will add a drive letter to the paths below since we
+	// canonicalize them.
+	if runtime.GOOS == "windows" {
+		return
+	}
+
+	path, err := ioutil.TempDir("", "lfsRepo")
+	assert.Nil(t, err)
+	path = path + "/local/path"
+	err = os.MkdirAll(path, 0755)
+	assert.Nil(t, err)
+
+	finder := NewEndpointFinder(lfshttp.NewContext(nil, nil, map[string]string{
+		"remote.origin.url": path + "/.git",
+	}))
+	e := finder.Endpoint("download", "")
+	assert.Equal(t, "file://"+path, e.Url)
+
+	os.RemoveAll(path)
 }
 
 func TestAccessConfig(t *testing.T) {

--- a/lfshttp/endpoint.go
+++ b/lfshttp/endpoint.go
@@ -3,6 +3,7 @@ package lfshttp
 import (
 	"fmt"
 	"net/url"
+	"os"
 	"path/filepath"
 	"regexp"
 	"strings"
@@ -108,9 +109,6 @@ func EndpointFromHttpUrl(u *url.URL) Endpoint {
 }
 
 func EndpointFromLocalPath(path string) Endpoint {
-	if !strings.HasSuffix(path, ".git") {
-		path = fmt.Sprintf("%s/.git", path)
-	}
 	var slash string
 	if abs, err := filepath.Abs(path); err == nil {
 		// Required for Windows paths to work.
@@ -118,6 +116,17 @@ func EndpointFromLocalPath(path string) Endpoint {
 			slash = "/"
 		}
 		path = abs
+	}
+
+	var gitpath string
+	if filepath.Base(path) == ".git" {
+		gitpath = path
+		path = filepath.Dir(path)
+	} else {
+		gitpath = filepath.Join(path, ".git")
+	}
+	if file, err := os.Lstat(gitpath); err == nil && file.IsDir() {
+		path = gitpath
 	}
 	return Endpoint{Url: fmt.Sprintf("file://%s%s", slash, filepath.ToSlash(path))}
 }

--- a/lfshttp/endpoint.go
+++ b/lfshttp/endpoint.go
@@ -125,7 +125,7 @@ func EndpointFromLocalPath(path string) Endpoint {
 	} else {
 		gitpath = filepath.Join(path, ".git")
 	}
-	if file, err := os.Lstat(gitpath); err == nil && file.IsDir() {
+	if _, err := os.Stat(gitpath); err == nil {
 		path = gitpath
 	}
 	return Endpoint{Url: fmt.Sprintf("file://%s%s", slash, filepath.ToSlash(path))}

--- a/lfshttp/standalone/standalone.go
+++ b/lfshttp/standalone/standalone.go
@@ -101,6 +101,11 @@ func gitDirAtPath(path string) (string, error) {
 	}
 	env = env[:n]
 
+	// Trim any trailing .git path segment.
+	if filepath.Base(path) == ".git" {
+		path = filepath.Dir(path)
+	}
+
 	curdir, err := os.Getwd()
 	if err != nil {
 		return "", err

--- a/t/cmd/util/testutils.go
+++ b/t/cmd/util/testutils.go
@@ -16,6 +16,7 @@ import (
 	"os"
 	"os/exec"
 	"path/filepath"
+	"runtime"
 	"strings"
 	"sync"
 	"time"
@@ -26,6 +27,34 @@ import (
 	"github.com/git-lfs/git-lfs/git"
 	"github.com/git-lfs/git-lfs/lfs"
 )
+
+func init() {
+	path := os.Getenv("PATH")
+	sep := ""
+	if path != "" {
+		if runtime.GOOS == "windows" {
+			sep = ";"
+		} else {
+			sep = ":"
+		}
+	}
+
+	// Strip the trailing "t/cmd/util/testutils.go" from the path to this
+	// source file to create a path to the working tree's "bin" directory,
+	// then prepend that to the PATH environment variable to ensure our
+	// "git-lfs" binary is used in preference to any installed versions when
+	// executing the Go tests.
+	_, srcdir, _, _ := runtime.Caller(0)
+	for i := 0; i < 4; i++ {
+		srcdir = filepath.Dir(srcdir)
+	}
+	var err error
+	srcdir, err = filepath.Abs(srcdir)
+	if err != nil {
+		panic(err)
+	}
+	os.Setenv("PATH", filepath.Join(srcdir, "bin")+sep+path)
+}
 
 type RepoType int
 

